fvm/evm/emulator/database/database.go

```diff
@@ -0,0 +1,341 @@
+	baseStorage := atree.NewLedgerBaseStorage(led)
+
+	storage, err := NewPersistentSlabStorage(baseStorage)
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	db := &Database{
+		led:                led,
+		flowEVMRootAddress: flowEVMRootAddress,
+		storage:            storage,
+	}
+
+	err = db.retrieveOrCreateMapRoot()
+	if err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+func (db *Database) retrieveOrCreateMapRoot() error {
+	rootIDBytes, err := db.led.GetValue(db.flowEVMRootAddress.Bytes(), []byte(FlowEVMRootSlabKey))
+	if err != nil {
+		return handleError(err)
+	}
+
+	var m *atree.OrderedMap
+	if len(rootIDBytes) == 0 {
+		m, err = atree.NewMap(db.storage, atree.Address(db.flowEVMRootAddress), atree.NewDefaultDigesterBuilder(), typeInfo{})
+		if err != nil {
+			return handleError(err)
+		}
+	} else {
+		storageID, err := atree.NewStorageIDFromRawBytes(rootIDBytes)
+		if err != nil {
+			return handleError(err)
+		}
+		m, err = atree.NewMapWithRootID(db.storage, storageID, atree.NewDefaultDigesterBuilder())
+		if err != nil {
+			return handleError(err)
+		}
+	}
+	db.atreemap = m
+	return nil
+}
+
+func (db *Database) storeMapRoot() error {
+	rootIDBytes := make([]byte, StorageIDSize)
+	_, err := db.atreemap.StorageID().ToRawBytes(rootIDBytes)
+	if err != nil {
+		return handleError(err)
+	}
+	err = db.led.SetValue(db.flowEVMRootAddress.Bytes(), []byte(FlowEVMRootSlabKey), rootIDBytes[:])
+	if err != nil {
+		return handleError(err)
+	}
+	return nil
+}
+
+// Get retrieves the given key if it's present in the key-value store.
+func (db *Database) Get(key []byte) ([]byte, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	return db.get(key)
+}
+
+func (db *Database) get(key []byte) ([]byte, error) {
+	data, err := db.atreemap.Get(compare, hashInputProvider, NewByteStringValue(key))
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	v, err := data.StoredValue(db.atreemap.Storage)
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	return v.(ByteStringValue).Bytes(), nil
+}
+
+// Put inserts the given value into the key-value store.
+func (db *Database) Put(key []byte, value []byte) error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+	return db.put(key, value)
+}
+
+func (db *Database) put(key []byte, value []byte) error {
+	_, err := db.atreemap.Set(compare, hashInputProvider, NewByteStringValue(key), NewByteStringValue(value))
+	return handleError(err)
+}
+
+// Has checks if a key is present in the key-value store.
+func (db *Database) Has(key []byte) (bool, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	data, err := db.get(key)
+	return len(data) > 0, err
+}
+
+// Delete removes the key from the key-value store.
+func (db *Database) Delete(key []byte) error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+	return db.delete(key)
+}
+
+func (db *Database) delete(key []byte) error {
+	_, _, err := db.atreemap.Remove(compare, hashInputProvider, NewByteStringValue(key))
+	return handleError(err)
```

Comment:
Atree `OrderedMap.Remove` returns removed key and value storables.  We probably want to remove those from storage if they are `StorageIDStorable`.  

```suggestion
	removedMapKeyStorable, removedMapValueStorable, err := db.atreemap.Remove(compare, hashInputProvider, NewByteStringValue(key))
	if err != nil {
		return handleError(err)
	}
	
	if id, ok := removedMapKeyStorable.(atree.StorageIDStorable); ok {
		// NOTE: deep remove isn't necessary because key is ByteStringValue (not container)
		err := db.storage.Remove(id)
	 	if err != nil {
	  		return handleError(err)
		}
	}	
	
	if id, ok := removedMapValueStorable.(atree.StorageIDStorable); ok {
		// NOTE: deep remove isn't necessary because value is ByteStringValue (not container)
		err := db.storage.Remove(id)
	 	if err != nil {
	  		return handleError(err)
		}
	}		
```